src/Components/Authorization/src/AttributeAuthorizeDataCache.cs

```diff
@@ -29,13 +29,22 @@ private static IAuthorizeData[] ComputeAuthorizeDataForType(Type type)
         {
             // Allow Anonymous skips all authorization
             var allAttributes = type.GetCustomAttributes(inherit: true);
-            if (allAttributes.OfType<IAllowAnonymous>().Any())
+            List<IAuthorizeData> authorizeDatas = null;
+            for (var i = 0; i < allAttributes.Length; i++)
```

Comment:
Lowering the abstraction from LINQ to a `for` loop is a bit of an odd way to move forwards. I could totally buy it if it was for runtime perf, but from what I understand, it's really only about reducing the post-linked bundle size. Is that right?

1. I'm not a fan of removing it just as a way to reduce the post-linked bundle size because it's deliberately missing the point of the whole size reduction exercise. Almost all customer apps are going to involve calls into `System.Linq` and typical methods like `.OfType` and `.Any`, so it's not really helping anyone to pretend we can trim them.
2. Even if we decide it's beneficial to remove the calls into `System.Linq`, we don't have to give up on the abstraction it offers. The implementations of `OfType` and `Any` (etc) are completely trivial - we could have our own micro-LINQ extension methods in our own code here, without costing anything meaningful in size. 

How best do you think the logic here can be expressed? If you think the `for` loop and chain of `if` blocks is clearer and more maintainable then great, let's do this, but if it's only a workaround for size concerns then let's discuss it further.

Reply:
Maybe in some cases the runtime perf benefits of lowering to the `for`+`if` type logic are meaningful in terms of being super clear that we aren't doing implicit allocations for iterators or capturing lambdas. For one-time-only initialization logic the benefit might be limited, but the core perf-critical stuff like rendering is written this way already.

I trust your judgement on whether each of the changes here are preferable versus having our own LINQ-like utilities or even continuing to use `System.Linq`. But in the cases where we're taking an existing very clear one-liner like "something?.OfType<...>().FirstOrDefault()` and converting it to 8 lines of procedural code with `continue` statements and null checks, it would be great to add a comment at the top summarizing what we're doing (e.g., `// Find the first blah`) as it greatly reduces the mental load on people skimming through the logic.

Reply:
OK. It wasn't any single change that I was questioning, but rather the net effect of all of these changes and the ongoing pattern of writing lower-level code even in non-perf-critical areas.

If you can at least put in the sort of comments that indicate, in a high-level way, what the chains of `for`/`if`/etc are doing (e.g., "Find the first X that has a Y") that would help a lot to retain the quick legibility. I know such comments come with a risk of getting stale, but don't know of a perfect alternative.

Reply:
On the other hand if you think there's a reasonable case for continuing to use `System.Linq` itself in non-perf-critical areas, I'd very likely get behind that, based on the reality that it's not going to be linked out from any realistic application anyway.

Reply:
I love this discussion 😄 . Also consider not using an array for the return here if this is purely an internal use case. Maybe Memory/Span would be more appropriate? That way you could use [CollectionsMarshal.AsSpan](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.asspan?view=net-5.0#System_Runtime_InteropServices_CollectionsMarshal_AsSpan__1_System_Collections_Generic_List___0__) to extract the underlying array if it's purely for enumeration.